Tests/PackageCollectionsSigningTests/SignatureTests.swift

```diff
@@ -20,149 +20,160 @@ import X509
-            let jsonEncoder = JSONEncoder()
-            let jsonDecoder = JSONDecoder()
-
-            let payload = ["foo": "bar"]
-
-            let certPath = fixturePath.appending(components: "Certificates", "Test_ec.cer")
-            let certData: Data = try localFileSystem.readFileContents(certPath)
-            let certificate = try Certificate(derEncoded: Array(certData))
-
-            // This is not cert's key so `parse` will fail
-            let privateKey = try P256.Signing.PrivateKey(pemRepresentation: ecPrivateKey)
-            let signature = try Signature.generate(
-                payload: payload,
-                certChainData: [certData],
-                jsonEncoder: jsonEncoder,
-                signatureAlgorithm: .ES256
-            ) {
-                try privateKey.signature(for: SHA256.hash(data: $0)).rawRepresentation
+    func test_ES256_generateAndValidate_keyMismatch() async throws {
+        let jsonEncoder = JSONEncoder.makeWithDefaults()
+        let jsonDecoder = JSONDecoder.makeWithDefaults()
+
+        let certData = try temp_await { callback in
+            self.readTestCertData(
+                path: { fixturePath in fixturePath.appending(components: "Certificates", "Test_ec.cer") },
+                callback: callback
+            )
+        }
+        let certificate = try Certificate(derEncoded: Array(certData))
+
+        let payload = ["foo": "bar"]
+        // This is not cert's key so `parse` will fail
+        let privateKey = try P256.Signing.PrivateKey(pemRepresentation: ecPrivateKey)
+        let signature = try Signature.generate(
+            payload: payload,
+            certChainData: [certData],
+            jsonEncoder: jsonEncoder,
+            signatureAlgorithm: .ES256
+        ) {
+            try privateKey.signature(for: SHA256.hash(data: $0)).rawRepresentation
+        }
+
+        do {
+            _ = try await Signature.parse(
+                signature,
+                certChainValidate: { _ in [certificate] },
+                jsonDecoder: jsonDecoder
+            )
+            XCTFail("Expected error")
+        } catch {
+            guard SignatureError.invalidSignature == error as? SignatureError else {
+                return XCTFail("Expected SignatureError.invalidSignature")
             }
+        }
+    }
 
-            XCTAssertThrowsError(try temp_await { callback in
-                Signature.parse(
-                    signature,
-                    certChainValidate: { _, cb in cb(.success([certificate])) },
-                    jsonDecoder: jsonDecoder,
-                    callback: callback
-                )
-            }) { error in
-                guard SignatureError.invalidSignature == error as? SignatureError else {
-                    return XCTFail("Expected SignatureError.invalidSignature")
-                }
+    private func readTestCertData(path: (AbsolutePath) -> AbsolutePath, callback: (Result<Data, Error>) -> Void) {
```

Comment:
Can this be made `async`? The tests calling it seem to be `async` already.

Reply:
amended https://github.com/apple/swift-package-manager/pull/6609/commits/813bd986f020f69b88b41826970a3223896516b7